core/controller/src/main/scala/whisk/core/controller/Meta.scala

```diff
@@ -16,23 +16,135 @@
 import whisk.core.entity._
 import whisk.core.entity.types._
 import whisk.http.ErrorResponse.terminate
 import whisk.http.Messages
+import whisk.utils.JsHelpers._
+
+private case class Context(
+    method: HttpMethod,
+    headers: List[HttpHeader],
+    path: String,
+    query: Map[String, String],
+    body: Option[JsObject] = None) {
+    val requestParams = query.toJson.asJsObject.fields ++ { body.map(_.fields) getOrElse Map() }
+    val overrides = WhiskMetaApi.reservedProperties.intersect(requestParams.keySet)
+    def withBody(b: Option[JsObject]) = Context(method, headers, path, query, b)
+    def metadata(user: Option[Identity]) = {
+        WhiskMetaApi.metadata(method.value, headers.mkString("\n"), path, user.map(_.namespace.asString))
+    }
+}
+
+object WhiskMetaApi extends Directives {
+    /** Reserved parameters that requests may no defined. */
+    val reservedProperties = Set(
+        "__ow_meta_verb",
+        "__ow_meta_headers",
+        "__ow_meta_path",
+        "__ow_meta_namespace")
+
+    /** Supported media types for action response. */
+    val mediaTranscoders = Map(
+        "html" -> resultAsHtml _,
+        "text" -> resultAsText _,
+        "js" -> resultAsJson _,
+        "http" -> resultAsHttp _)
+
+    val supportedMediaTypes = mediaTranscoders.keySet
+
+    def metadata(method: String, headers: String, path: String, namespace: Option[String]) = {
+        Map("__ow_meta_verb" -> method.toLowerCase.toJson,
+            "__ow_meta_headers" -> headers.toJson,
+            "__ow_meta_path" -> path.toJson) ++
+            namespace.map(ns => "__ow_meta_namespace" -> ns.toJson)
+    }
+
+    protected def resultAsHtml(result: JsValue): RequestContext => Unit = {
+        Try {
+            val JsString(str) = result
+            val input = new ByteArrayInputStream(str.getBytes)
+            scala.xml.XML.load(input)
+        } match {
+            case Success(xml) => respondWithMediaType(`text/html`) { complete(OK, xml) }
+            case Failure(t)   => complete(BadRequest, Messages.invalidMedia(`text/html`))
+        }
+    }
+
+    protected def resultAsText(result: JsValue): RequestContext => Unit = {
+        result match {
+            case r: JsObject  => complete(OK, r.compactPrint)
+            case r: JsArray   => complete(OK, r.compactPrint)
+            case JsString(s)  => complete(OK, s)
+            case JsBoolean(b) => complete(OK, b.toString)
+            case JsNumber(n)  => complete(OK, n.toString)
+            case JsNull       => complete(OK)
+        }
+    }
+
+    protected def resultAsJson(result: JsValue): RequestContext => Unit = {
+        result match {
+            case r: JsObject  => complete(OK, r)
+            case r: JsArray   => complete(OK, r)
+            case JsString(s)  => complete(OK, s)
+            case JsBoolean(b) => complete(OK, b.toString)
+            case JsNumber(n)  => complete(OK, n.toString)
+            case JsNull       => complete(OK)
```

Comment:
A proper JSON response is indeed either an array or object, not a primitive type.
So the marshaler could treat these as error cases instead and if the caller desired to serialize primitive values then they must request `text` as the extension.